s3tests_boto3/functional/test_s3select.py

```diff
@@ -0,0 +1,410 @@
+
+    # assert is according to radom-csv function 
+    assert res_s3select == "0,1000,1,"
+
+    # purpose of test that all where conditions create the same group of values, thus same result
+    res_s3select_substr = remove_xml_tags_from_result(  run_s3select(bucket_name,csv_obj_name,'select min(int(_2)),max(int(_2)) from stdin where substr(_2,1,1) == "1"')).replace("\n","")
+
+    res_s3select_between_numbers = remove_xml_tags_from_result(  run_s3select(bucket_name,csv_obj_name,'select min(int(_2)),max(int(_2)) from stdin where int(_2)>=100 and int(_2)<200')).replace("\n","")
+
+    res_s3select_eq_modolu = remove_xml_tags_from_result(  run_s3select(bucket_name,csv_obj_name,'select min(int(_2)),max(int(_2)) from stdin where int(_2)/100 == 1 or int(_2)/10 == 1 or int(_2) == 1')).replace("\n","")
+
+    assert res_s3select_substr == res_s3select_between_numbers
+
+    assert res_s3select_between_numbers == res_s3select_eq_modolu
+    
+def test_alias():
+
+    # purpose: test is comparing result of exactly the same queries , one with alias the other without.
+    # this test is setting alias on 3 projections, the third projection is using other projection alias, also the where clause is using aliases
+    # the test validate that where-clause and projections are executing aliases correctly, bare in mind that each alias has its own cache,
+    # and that cache need to be invalidate per new row. 
+
+    csv_obj = create_random_csv_object(10000,10)
+
+    csv_obj_name = "csv_10000x10"
+    bucket_name = "test"
+    upload_csv_object(bucket_name,csv_obj_name,csv_obj)
+
+    res_s3select_alias = remove_xml_tags_from_result(  run_s3select(bucket_name,csv_obj_name,"select int(_1) as a1, int(_2) as a2 , (a1+a2) as a3 from stdin where a3>100 and a3<300;")  ).replace(",","")
+
+    res_s3select_no_alias = remove_xml_tags_from_result(  run_s3select(bucket_name,csv_obj_name,"select int(_1),int(_2),int(_1)+int(_2) from stdin where (int(_1)+int(_2))>100 and (int(_1)+int(_2))<300;")  ).replace(",","")
+
+    assert res_s3select_alias == res_s3select_no_alias
+
+
+def test_alias_cyclic_refernce():
+
+    # purpose of test is to validate the s3select-engine is able to detect a cyclic reference to alias.
+    
+    csv_obj = create_random_csv_object(10000,10)
+
+    csv_obj_name = "csv_10000x10"
+    bucket_name = "test"
+    upload_csv_object(bucket_name,csv_obj_name,csv_obj)
+
+    res_s3select_alias = remove_xml_tags_from_result(  run_s3select(bucket_name,csv_obj_name,"select int(_1) as a1,int(_2) as a2, a1+a4 as a3, a5+a1 as a4, int(_3)+a3 as a5 from stdin;")  )
+
+    find_res = res_s3select_alias.find("number of calls exceed maximum size, probably a cyclic reference to alias");
+    
+    assert int(find_res) >= 0 
+
+def test_datetime():
+
+    # purpose of test is to validate date-time functionality is correct,
```

Comment:
random tests are great! but, in addition, we can add more accurate tests, where the input csv is not randomly generated, and the result set is checked against the expected result

Reply:
generally many tests are missing, there is no full coverage.

s3-select functionalities is quite complex, each function may contain several flows(including error-flows)
thus,accurate test with accurate result, should added.

please note, that with random input (datetime in this case) , combine with complex query (datediff,dateadd,date-extract,substr), actually validate result correctness.